sw/device/lib/dif/dif_adc_ctrl.h

```diff
@@ -99,6 +99,13 @@ typedef enum dif_adc_ctrl_irq_cause {
    * Sample ready cause in Oneshot mode.
    */
   kDifAdcCtrlIrqCauseOneshot = 1U << 8,
```

Comment:
Since this enum is used in a way that requires the values to be the same as the CSR, I'd recommend using the constant from adc_ctrl_regs.h, instead of the hard-coded `8`. This value could move. (Same for `kDifAdcCtrlIrq`)

Reply:
So this is how I wanted to do it originally, but I ran into a meson issue. Specifically, autogenerated headers are listed as a ["source"](https://github.com/lowRISC/opentitan/blob/e7149d214a607b92e1d31591246f4c6f7dff7962/sw/device/lib/dif/meson.build#L22) in build targets, taking the [advice from meson docs](https://mesonbuild.com/Generating-sources.html#generating-headers). Since the register headers are autogenerated on the fly, rather than checked in to the repo like the autogenerated DIF headers, to include it in the DIF header would require making it a dependency of the build target. Unfortunately, we generate these using a [meson "generator"](https://github.com/lowRISC/opentitan/blob/e7149d214a607b92e1d31591246f4c6f7dff7962/meson.build#L207),  which throws an error `ERROR: Argument is of an unacceptable type 'GeneratedList'.` when I try to include the `hw_ip_adc_ctrl_reg_h` in the [`dependencies` list](https://github.com/lowRISC/opentitan/blob/e7149d214a607b92e1d31591246f4c6f7dff7962/sw/device/lib/dif/meson.build#L25). 

One work-around I encountered was to list `hw_ip_adc_ctrl_reg_h` as a source for _every_ meson build rule that includes the  `sw_lib_dif_adc_ctrl` lib, but I thought this was annoying, and since it was inconsistent with other DIF libraries, I opted against it. Perhaps there is another workaround I am not aware of? I thought it would be more wise to just file an issue on this and make it something we fix when the DIFs are built solely by bazel? Wdyt?

Reply:
Filing an issue and fixing with bazel sounds like a good path forward. Thanks!

Reply:
> autogenerated headers are listed as a ["source"](https://github.com/lowRISC/opentitan/blob/e7149d214a607b92e1d31591246f4c6f7dff7962/sw/device/lib/dif/meson.build#L22) in build targets

> Unfortunately, we generate these using a [meson "generator"](https://github.com/lowRISC/opentitan/blob/e7149d214a607b92e1d31591246f4c6f7dff7962/meson.build#L207), which throws an error ERROR: Argument is of an unacceptable type 'GeneratedList'. when I try to include the hw_ip_adc_ctrl_reg_h in the [dependencies list](https://github.com/lowRISC/opentitan/blob/e7149d214a607b92e1d31591246f4c6f7dff7962/sw/device/lib/dif/meson.build#L25).

So... are you using it as a source or as a pkg-config dependency? Because it sounds like you are trying to do both. :)

Reply:
What I would like to do is the latter, but the header is [autogenerated](https://github.com/lowRISC/opentitan/blob/5abad2426ef3579f59532be7a0734507d60edd96/meson.build#L207) and meson won't let me list this target as a dependency. Right now, the autogenerated header is listed as a "source" file, which works, but only when [the autogenerated header is listed as a source file in the same "sources" list](https://github.com/lowRISC/opentitan/blob/5abad2426ef3579f59532be7a0734507d60edd96/sw/device/lib/dif/meson.build#L22), NOT when it is included in a header that is checked into our repo, that itself is included in a source file in the sources list. (i.e., unlike bazel, meson seems to not explicitly track header files, so it seems it has no way of knowing that the autogeneration of a header file is a dependency for a manually implemented header?)

Reply:
Meson does track headers. It does so in the `sources` kwarg as you mentioned above -- any of those sources that are autogenerated and aren't .c or .cpp or some other file that gets directly compiled, get added as "header deps" which means that they must be generated first thing, before any .c or .cpp files in that library/executable get compiled.

`dependencies` is for adding the CFLAGS and LIBS kind of dependencies, it's not for adding the "must be built first" kind of dependencies.

Could you explain what problem you had when you tried to use the autogenerated header in `sources`?

Reply:
The below setup yields the error:
"fatal error: 'adc_ctrl_regs.h' file not found" (**when including the sw_lib_dif_adc_ctrl lib in the deps list of other rules**)

manually implemented header (`dif_adc_ctrl.h`):
```c
#include "adc_ctrl_regs.h" // Generated

// use some macros in autogen header
```

source file (`dif_adc_ctrl.c`):
```c
#include "adc_ctrl_regs.h" // Generated
#include "dif_adc_ctrl.h"
```

meson build rule, where \<autogen header target\> = "adc_ctrl_regs.h"
```
sw_lib_dif_adc_ctrl = declare_dependency(
    link_with: static_library(
        'sw_lib_dif_adc_ctrl',
         sources: [
            <autogen header target>,
            'dif_adc_ctrl.c',
         ],
         dependencies: [
            ...
         ],
    )
)
```
-------------
And using the same setup described above, with the slightly different meson rule below setup yields the error:
"ERROR: Argument is of an unacceptable type 'GeneratedList'"

different meson build rule, where \<autogen header target\> = "adc_ctrl_regs.h"
```
sw_lib_dif_adc_ctrl = declare_dependency(
    link_with: static_library(
        'sw_lib_dif_adc_ctrl',
         sources: [
            'dif_adc_ctrl.c',
         ],
         dependencies: [
            <autogen header target>,
            ...
         ],
    )
)
```